### PR TITLE
Attribute eth_call RPC usage by consumer

### PR DESCRIPTION
**Pull request by Jurij89**
## Summary

This chained follow-up into PR #1440 adds attribution for the remaining `eth_call` startup noise called out in #1480 without changing node behavior.

High-level intent:
- keep the existing aggregate `rpc_usage method=<method>` log contract unchanged;
- propagate stable code-owned read labels through adapter-owned failover reads using async-local context;
- count raw `eth_call` attempts by consumer at the provider billing boundary, including retry/failover attempts;
- emit companion `rpc_usage_by_consumer method=eth_call consumer=<label> ...` daemon log lines so the next mainnet-base startup run can identify the largest consumers before we optimize them.

This is deliberately attribution-first. It does not cache, delay, throttle, or otherwise alter chain reads.

## Related Work

- Follow-up to PR #1440 (`codex/rpc-call-reduction`)
- Related to issue #1480 and the fresh `mainnet-base` feedback showing first-minute `eth_call` around `713/min` after the previous `eth_getLogs` / `eth_chainId` reductions.

## Files Changed

- `packages/chain/src/rpc-usage.ts`: adds bounded consumer labels, async-local scope helpers, optional `ethCallByConsumer` windows, merge support, and raw provider/retry attribution through a single tracker-owned recording boundary.
- `packages/chain/src/rpc-failover-client.ts`: scopes adapter-owned `read` and `readContract` operations with the existing logical read label.
- `packages/cli/src/daemon/rpc-usage-log.ts`: preserves aggregate `rpc_usage` lines and adds companion `rpc_usage_by_consumer` lines for `eth_call`.
- Tests cover tracker behavior, async scope isolation, provider and contract-view failover attribution, formatter output, and the agent telemetry boundary.

## Test Plan

- [x] `pnpm --dir packages/chain exec vitest run --config vitest.unit.config.ts test/rpc-usage.unit.test.ts --reporter dot`
- [x] `pnpm --dir packages/cli exec vitest run --config vitest.unit.config.ts test/rpc-usage-log.test.ts --reporter dot`
- [x] `pnpm --dir packages/agent exec vitest run test/rpc-usage-boundary.test.ts --reporter dot`
- [x] `pnpm --dir packages/chain run build`
- [x] `pnpm --dir packages/chain exec tsc --noEmit`
- [x] `pnpm --dir packages/cli exec tsc --noEmit`
- [x] `pnpm --dir packages/agent exec tsc --noEmit`
- [x] `git diff --check`

## Notes

Local focused chain tests rewrote `packages/evm-module/deployments/localhost_contracts.json`; that generated deployment artifact is intentionally left out of this PR.

### CODE REVIEW

File: packages/chain/src/rpc-usage.ts [hunk 1]
```diff
@@ -180,14 +226,25 @@ export class RpcUsageTracker {
    * one rpc_usage log line per fabricated name every minute.
    */
   static readonly MAX_WINDOW_METHODS = 64;
+  static readonly MAX_WINDOW_CONSUMERS = 128;
 
-  record(method: string): void {
+  record(method: string, consumer = currentRpcUsageConsumer()): void {
```
Comment (otReviewAgent) [inline packages/chain/src/rpc-usage.ts]: 🟡 Issue: Make RPC consumer attribution a single explicit boundary instead of mixing ambient context with optional arguments

**What's wrong**
The change makes the tracker’s behavior depend on non-local ambient state while also exposing an explicit parameter for the same data. That forces readers to hold two attribution mechanisms in their head and makes future accounting call sites easy to get subtly inconsistent. The new eth_call-only branch also lands directly in the generic request-counting hot path, so the tracker is starting to accumulate diagnostic feature logic instead of keeping a small, obvious model.

**Example**
A future call site reading `tracker.record('eth_call')` cannot tell from the call whether it updates only `byMethod` or also `ethCallByConsumer`; that depends on an outer `withRpcUsageConsumer(...)` scope that may be several layers away. The provider factory also passes `currentRpcUsageConsumer()` even though `record()` already has the same default, which shows the attribution boundary is not cleanly owned by either caller or tracker.

**Suggested direction**
Collapse this into one owner for consumer attribution. If AsyncLocalStorage is still the right mechanism for async propagation, hide it behind a narrow tracker/scope API and let the recording path consume it in one place. If the provider callback should own the label, make that explicit and remove the default ambient lookup. Either way, avoid a public `currentRpcUsageConsumer()` plus an optional `consumer` parameter plus caller-side explicit reads of the same ambient state.

**Confidence note**
This is a maintainability concern rather than a behavior claim; the current behavior can be preserved while tightening the boundary.

**For Agents**
In packages/chain/src/rpc-usage.ts and packages/chain/src/rpc-failover-client.ts, preserve aggregate counts and eth_call consumer attribution, but make attribution a single explicit abstraction. Prefer either an explicit `record(method, { consumer })`/callback contract or a tracker-owned `withConsumer(label, fn)` scope with no exported ambient getter and no duplicate caller-side `currentRpcUsageConsumer()` calls. Keep coverage for overlapping async scopes and failover retry attempts inheriting the same label.

File: packages/chain/src/rpc-failover-client.ts [hunk 2]
```diff
@@ -224,30 +228,33 @@ export class RpcFailoverClient {
...
-          this.recordRpcOutcome('eth_call', this.rpcOutcome(err), { retryable: isRetryableRpcError(err) });
-          throw err;
-        } finally {
-          metrics.chainRpcDuration.record(Date.now() - startedAt, {
-            rpc_method: 'eth_call', chain_id: chainId,
-          });
-        }
-      },
-      { attributes: { 'rpc.method': 'eth_call', 'dkg.chain_id': chainId, 'dkg.read': label } },
+    return withRpcUsageConsumer(
```
Comment (otReviewAgent) [inline packages/chain/src/rpc-failover-client.ts]: 🟡 Issue: Contract-view eth_call attribution is not verified

**What's wrong**
Most eth_call consumers in this codebase go through readContract, and this diff adds a separate wrapper there. The added end-to-end attribution test uses readProvider, so it does not prove that contract rebind, withSpan, and the contract-view failover path preserve the consumer label into the transport accounting.

**Example**
A regression that removed only the readContract wrapper would still let the new tracker tests pass and would still let the readProvider failover attribution test pass, while real contract view reads such as token.allowance or pcaNFT.getAccountInfo would drain with byMethod.eth_call but no ethCallByConsumer entry for the read label.

**Suggested direction**
Cover the readContract path explicitly, not just readProvider and the standalone AsyncLocalStorage helper.

**For Agents**
Add an integration-style test around EVMChainAdapter.readContract or RpcFailoverClient.readContract with a usage-counted provider. Trigger a real contract-view eth_call, drain RPC usage, and assert both usage.byMethod.eth_call and usage.ethCallByConsumer[label] match the raw eth_call attempts, including a failover or retry case if feasible.

File: packages/chain/src/rpc-usage.ts (hunk 1 above)
Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts]: Addressed in 4df06f8ee. `RpcUsageTracker.record(method)` is now the single recording boundary: the ambient getter is no longer exported, the optional consumer argument is gone, and provider/retry hooks no longer read the context themselves. The existing async-scope and failover retry coverage still passes with `pnpm --dir packages/chain exec vitest run --config vitest.unit.config.ts test/rpc-usage.unit.test.ts --reporter dot`.

File: packages/chain/src/rpc-failover-client.ts (hunk 2 above)
Comment (Jurij89) [reply packages/chain/src/rpc-failover-client.ts]: Addressed in 4df06f8ee. Added `attributes failover contract-view eth_call attempts to the readContract label`, which drives a real ethers Contract view through `EVMChainAdapter.readContract` with a throttled primary and healthy backup, then asserts both aggregate `byMethod.eth_call` and `ethCallByConsumer[unit.contract.value]` match the raw endpoint hits. Verified with `pnpm --dir packages/chain exec vitest run --config vitest.unit.config.ts test/rpc-usage.unit.test.ts --reporter dot`.

File: packages/chain/src/rpc-usage.ts (hunk 1 above)
Comment (otReviewAgent) [reply packages/chain/src/rpc-usage.ts]: 🟡 Issue: Avoid making the concrete usage window grow an optional companion shape

**What's wrong**
`RpcUsageWindow` was intentionally a concrete drain contract, but the new optional map makes the new dimension a partial/ambient concern. That spreads the same absence handling through merge, drain, formatter, and tests, and makes future drain sources easier to get subtly out of sync.

**Example**
A new `RpcUsageDrainable` can now return `{ byMethod: { eth_call: 10 }, lifetimeTotal: 10 }` and still satisfy the type even if it internally tracks consumer attribution but forgot to expose it. Every merge/format caller must defensively treat missing and empty as equivalent instead of the window model making that invariant obvious.

**Suggested direction**
Prefer a single normalized window shape, e.g. `ethCallByConsumer: Record<string, number>` with `{}` as the identity, or centralize construction through helpers that normalize this field before windows cross the drain boundary. That would delete the repeated `?? {}` / conditional object-shape branches and keep the “always concrete window” contract intact.

**Confidence note**
The optional shape may be intentional for backward-compatible object output, but within this PR it reintroduces absence-vs-empty branching into a model that was previously explicit about collapsing optionality at the boundary.

**For Agents**
Look at `RpcUsageWindow`, `emptyRpcUsageWindow`, `RpcUsageTracker.drainWindow`, `mergeRpcUsageWindows`, mock adapter windows, and `formatRpcUsageLines`. Preserve the existing aggregate `rpc_usage` output and only emit consumer log lines for nonzero entries. Add or adjust a focused unit test proving an empty consumer map produces no companion lines while merge/drain always return a single normalized window shape.

File: packages/chain/src/rpc-failover-client.ts (hunk 2 above)
Comment (otReviewAgent) [reply packages/chain/src/rpc-failover-client.ts]: 🟡 Issue: Consumer attribution lacks coverage for same-endpoint retry attempts

**What's wrong**
The new tests verify attribution across failover endpoints, but not the retry path used by single-RPC nodes. That path records additional raw requests below the initial provider send, and losing the consumer context there would under-report `ethCallByConsumer` while preserving the existing aggregate request count.

**Example**
A single-RPC adapter with `startLoopbackRpc({ throttle: ['eth_call'] })` can make `a.readProvider('unit.eth_call.retry', p => p.send('eth_call', ...))` reject after retries. The regression test should then assert `usage.byMethod.eth_call === rpc.hits('eth_call')` and `usage.ethCallByConsumer['unit.eth_call.retry'] === rpc.hits('eth_call')`. Without that, attribution can silently miss retry-hook counts while the current failover tests still pass.

**Suggested direction**
Cover the single-RPC retry path separately from multi-endpoint failover so the AsyncLocalStorage consumer context is proven to reach counts recorded by the retry hook.

**For Agents**
Add a single-endpoint `eth_call` retry regression test near the new RPC usage attribution tests. Use the loopback throttle harness, drain usage after the rejected read, and prove both aggregate `byMethod.eth_call` and `ethCallByConsumer[label]` match the server hit count.

File: packages/chain/src/rpc-usage.ts (hunk 1 above)
Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts]: Addressed in ad6ff9c64. `RpcUsageWindow` now has a normalized concrete `ethCallByConsumer: {}` identity, `emptyRpcUsageWindow()`, `RpcUsageTracker.drainWindow()`, `MockChainAdapter`, and `mergeRpcUsageWindows()` all return the same shape, and the formatter reads the concrete map while still emitting no companion lines for an empty map. Verified with chain/CLI/agent focused tests, chain build, CLI/agent type checks, and `git diff --check`.

File: packages/chain/src/rpc-failover-client.ts (hunk 2 above)
Comment (Jurij89) [reply packages/chain/src/rpc-failover-client.ts]: Addressed in ad6ff9c64. Added `attributes same-endpoint eth_call retry attempts to the readProvider label`, which uses a single throttled loopback endpoint and asserts both aggregate `byMethod.eth_call` and `ethCallByConsumer[unit.eth_call.retry]` match the raw server hit count after bounded retries. Verified with `pnpm --dir packages/chain exec vitest run --config vitest.unit.config.ts test/rpc-usage.unit.test.ts --reporter dot`.

File: packages/chain/src/rpc-usage.ts [hunk 3]
```diff
@@ -120,12 +121,16 @@ export function mergeRpcUsageWindows(
   const defined = windows.filter((w): w is RpcUsageWindow => w !== undefined);
   if (defined.length === 0) return emptyRpcUsageWindow();
   const byMethod: Record<string, number> = {};
+  const ethCallByConsumer: Record<string, number> = {};
   let lifetimeTotal = 0;
   for (const w of defined) {
     for (const [m, c] of Object.entries(w.byMethod)) byMethod[m] = (byMethod[m] ?? 0) + c;
+    for (const [consumer, c] of Object.entries(w.ethCallByConsumer)) {
```
Comment (otReviewAgent) [inline packages/chain/src/rpc-usage.ts:138]: 🔴 Bug: New attribution field breaks existing RPC usage drain windows

**What's wrong**
The PR adds `ethCallByConsumer` as a required field and immediately iterates it. Any existing or out-of-tree `RpcUsageDrainable` implementation that still returns the previous `{ byMethod, lifetimeTotal }` shape now crashes the merge/format path. Because the daemon catches telemetry errors broadly, this can silently suppress all RPC usage logs for that window, not just the new consumer breakdown.

**Example**
A drain source that was valid before this PR, such as `{ byMethod: { eth_call: 2 }, lifetimeTotal: 2 }`, now makes `mergeRpcUsageWindows(...)` throw `TypeError: Cannot convert undefined or null to object`. In the daemon emission path, that exception is swallowed by `emitRpcUsage`, so even the aggregate `rpc_usage method=eth_call count=2 ...` line is lost.

**Suggested direction**
Default missing `ethCallByConsumer` to an empty object at merge/format boundaries, or introduce a normalization helper so the new diagnostic dimension does not invalidate the previous public drain contract.

**For Agents**
In `packages/chain/src/rpc-usage.ts` and `packages/cli/src/daemon/rpc-usage-log.ts`, preserve compatibility with windows that lack `ethCallByConsumer` by treating it as `{}` or by making the field optional at the boundary. Add a regression case proving legacy `{byMethod,lifetimeTotal}` windows still merge/emit aggregate `rpc_usage` lines while simply omitting consumer attribution.

Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts:138]: Addressed in 1822d6da7. Local drains still return a normalized `ethCallByConsumer: {}` map, but `mergeRpcUsageWindows()` and `formatRpcUsageLines()` now treat legacy windows without that field as empty attribution rather than throwing, so aggregate `rpc_usage` lines are preserved. Added regressions for legacy merge and legacy formatter input. Verified with the focused chain/CLI/agent RPC tests, chain build/tsc, CLI/agent tsc, and `git diff --check`.

Comment (otReviewAgent) [reply packages/chain/src/rpc-usage.ts:138]: 🟡 Issue: Keep the drained usage window normalized instead of optional

**What's wrong**
This turns a concrete telemetry model into a partially-normalized one. The new field is treated as optional everywhere even though all chain-owned producers return a concrete map, so compatibility concerns leak into the primary type and future consumers have to rediscover the `missing means empty` rule.

**Example**
A new formatter or merger can now receive `{ byMethod: { eth_call: 1 }, lifetimeTotal: 1 }` and must remember that missing `ethCallByConsumer` means the same thing as an empty map. The core model no longer makes that invariant explicit.

**Suggested direction**
Make `ethCallByConsumer` required on the normalized window model. If backwards compatibility with pre-attribution drain sources is needed, isolate that compatibility in one normalizer used by merge/format/drain boundaries instead of leaking optionality into every consumer.

**For Agents**
In `packages/chain/src/rpc-usage.ts`, keep the internal/public normalized `RpcUsageWindow` shape concrete, and introduce a separate loose/legacy input type or `normalizeRpcUsageWindow()` at compatibility boundaries. Preserve legacy drain-source tolerance, and prove legacy windows still emit aggregate lines while normalized windows always expose `ethCallByConsumer: {}`.

File: packages/chain/src/rpc-failover-client.ts [hunk 4]
```diff
@@ -196,11 +197,14 @@ export class RpcFailoverClient {
     fn: (provider: JsonRpcProvider) => Promise<T>,
     opts?: ReadOpts,
   ): Promise<T> {
-    return this.runAcrossProviders(
+    return withRpcUsageConsumer(
```
Comment (otReviewAgent) [inline packages/chain/src/rpc-failover-client.ts]: 🟡 Issue: Do not overload the read label as the consumer metric key

**What's wrong**
The `label` parameter previously described the read for failover/span context. This change gives it a hidden second responsibility: it now defines a bounded, logfmt-safe, stable telemetry dimension through `AsyncLocalStorage`. That makes every current and future read label part of the daemon logging contract without a type or API boundary saying so.

**Example**
Existing labels such as `hasContractCode getCode` and dynamic labels like ``pca rpc ${method}`` are now telemetry consumer keys after sanitization. A future edit that changes a label for human log clarity, or adds an address/block to aid debugging, would silently change the emitted metric dimension or collapse it to `other`.

**Suggested direction**
Separate the human/failover/span label from the bounded telemetry consumer key. A small typed model or explicit option would make the low-cardinality, stable log key intentional instead of ambient.

**Confidence note**
This is a maintainability finding: the current behavior may be intentional, but the diff makes an existing diagnostic label carry a second, stricter telemetry contract without making that boundary explicit.

**For Agents**
Look at `packages/chain/src/rpc-failover-client.ts` and the adapter read facades in `evm-adapter-base.ts`. Preserve the current emitted consumer names where they are desired, but introduce an explicit `rpcUsageConsumer` key, typed label object, or small set of constants so display/failover labels can evolve independently. Add a focused test showing a human label can contain spaces/change wording without changing the consumer key.

File: packages/chain/src/rpc-usage.ts (hunk 3 above)
Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts:138]: Addressed in f462fe692. `RpcUsageWindow.ethCallByConsumer` is required/concrete again; legacy missing-field tolerance now lives in the separate `RpcUsageWindowInput` + `normalizeRpcUsageWindow()` compatibility boundary used by merge/daemon formatting. Added regression coverage proving legacy `{ byMethod, lifetimeTotal }` inputs still normalize to `{ ethCallByConsumer: {} }` and preserve aggregate `rpc_usage` lines.

Verified with focused chain RPC usage tests, chain build/tsc, CLI formatter tests after rebuilding chain, CLI/agent tsc, agent boundary tests, failover unit tests, and `git diff --check`.

File: packages/chain/src/rpc-failover-client.ts (hunk 4 above)
Comment (Jurij89) [reply packages/chain/src/rpc-failover-client.ts]: Addressed in f462fe692. `RpcFailoverClient` now has an explicit `ReadOpts.rpcUsageConsumer` for the stable low-cardinality `eth_call` attribution key; the human failover/span `label` is no longer implicitly used as the metric key. Adapter read facades preserve today's emitted consumer names by passing that option explicitly, and the new regression proves a human label with spaces can change without changing the consumer key.

Verified with focused chain RPC usage tests, chain failover unit tests, chain build/tsc, CLI formatter tests, CLI/agent tsc, agent boundary tests, and `git diff --check`.

File: packages/chain/src/rpc-usage.ts (hunk 3 above)
Comment (otReviewAgent) [reply packages/chain/src/rpc-usage.ts:138]: 🔴 Bug: RpcUsageWindow makes a rollout-compatible field a source-breaking requirement

**What's wrong**
The implementation intentionally tolerates pre-attribution windows at runtime through RpcUsageWindowInput and normalizeRpcUsageWindow, but the exported RpcUsageWindow interface now requires ethCallByConsumer. That makes existing third-party or package-external ChainAdapter/RpcUsageDrainable implementations fail TypeScript compilation even though their runtime output is still supported.

**Example**
An existing adapter compiled against the public ChainAdapter contract can still return the legacy runtime shape that this PR explicitly normalizes elsewhere:

class ExternalAdapter implements ChainAdapter {
  drainRpcUsage() {
    return { byMethod: { eth_call: 1 }, lifetimeTotal: 1 };
  }
}

After this change, that implementation no longer type-checks because ethCallByConsumer is required, even though formatRpcUsageLines/mergeRpcUsageWindows accept and normalize the same legacy shape at runtime.

**Suggested direction**
Use the loose compatibility shape at public drain boundaries, or make the new field optional on the exported public window type and normalize before consumption. If the concrete internal model must require ethCallByConsumer, expose it under a separate internal/concrete type rather than changing the existing public contract in place.

**For Agents**
Look at packages/chain/src/rpc-usage.ts and packages/chain/src/chain-adapter.ts. Preserve runtime compatibility for legacy windows and also preserve source compatibility for external ChainAdapter/RpcUsageDrainable implementors. A fix should prove that a drain source returning { byMethod, lifetimeTotal } still type-checks at the public boundary and is normalized before formatting/merging.

Comment (otReviewAgent) [reply packages/chain/src/rpc-usage.ts:138]: 🟡 Issue: Legacy window compatibility is leaking into the canonical telemetry model

**What's wrong**
This adds optionality to the core telemetry abstraction instead of isolating it at the compatibility boundary. The PR now has both `RpcUsageWindow` and `RpcUsageWindowInput`, normalizes inside pure merge logic, and exports the loose shape publicly. That makes the real invariant harder to reason about and invites new in-process producers to keep omitting `ethCallByConsumer` instead of conforming to the concrete model.

**Example**
`mergeRpcUsageWindows({ byMethod: { eth_call: 1 }, lifetimeTotal: 1 })` is now a valid chain-side model input even though package-owned producers are supposed to return a concrete `RpcUsageWindow` with `ethCallByConsumer`.

**Suggested direction**
Move mixed-version/legacy tolerance to the CLI or drain edge, and keep the chain telemetry model strict internally and publicly.

**For Agents**
Keep `RpcUsageWindow` as the strict chain-owned model and keep `mergeRpcUsageWindows` typed to `RpcUsageWindow | undefined`. If the daemon formatter must tolerate old drain sources, define a local legacy input type and normalization edge in `packages/cli/src/daemon/rpc-usage-log.ts`; avoid exporting the loose input/normalizer from `packages/chain/src/index.ts`. Existing aggregate-only legacy formatting behavior should stay covered.

Comment (otReviewAgent) [reply packages/chain/src/rpc-usage.ts:138]: 🟡 Issue: Consumer attribution reset is not covered by a regression test

**What's wrong**
The changed behavior adds a second delta map that the daemon relies on for per-window telemetry. The tests prove attribution is recorded, but they do not prove it is drained as a delta. If this clear step regresses, stale consumer counts can leak into later emitted windows and overstate attribution.

**Example**
A regression test could record `eth_call` under `withRpcUsageConsumer('pcaNFT.getAccountInfo', ...)`, drain once and see the consumer count, then record only `eth_getLogs` or drain again and assert `ethCallByConsumer` is `{}`. Without clearing the consumer map, stale attribution could be emitted in a later non-empty window.

**Suggested direction**
Add a focused drain-reset assertion for `ethCallByConsumer`, preferably with a second window that contains non-`eth_call` traffic so stale consumer counts cannot hide behind an empty aggregate window.

**For Agents**
In `packages/chain/test/rpc-usage.unit.test.ts`, extend the new attribution coverage around `RpcUsageTracker.drainWindow()` to assert `ethCallByConsumer` is reset between drain windows while `lifetimeTotal` remains monotonic. Preserve existing aggregate `byMethod` delta semantics.

File: packages/chain/src/evm-adapter-base.ts [hunk 5]
```diff
@@ -981,7 +986,10 @@ export class EVMChainAdapterBase {
     fn: (provider: JsonRpcProvider) => Promise<T>,
     opts?: ReadOpts,
   ): Promise<T> {
-    return this.rpcFailover.read(label, fn, opts);
+    return this.rpcFailover.read(label, fn, {
```
Comment (otReviewAgent) [inline packages/chain/src/evm-adapter-base.ts:989]: 🟡 Issue: RPC usage attribution policy is split across two layers

**What's wrong**
The implementation makes `rpcUsageConsumer` optional in the failover client, then silently defaults it from `label` only in selected adapter helper methods. That spreads one telemetry policy across two abstractions and conflates a human failover/span label with a stable low-cardinality log key. The next maintainer can easily add a read at the failover layer and unknowingly skip attribution, or rename a human label and accidentally change the daemon log contract.

**Example**
A read through `readProvider('getNetwork (chainId)', ...)` gets a consumer because the adapter helper injects the label, while a direct `this.rpcFailover.read('validate configured chainId', ...)` call has the same kind of failover label but bypasses that defaulting policy. A future caller has to know which layer owns attribution.

**Suggested direction**
Push the consumer contract into the failover read boundary as a typed descriptor or required/explicit field, so labels, span names, and usage consumers have one clear ownership point.

**Confidence note**
The exact set of direct `rpcFailover.read` call sites may be intentional, but the changed structure makes that distinction implicit rather than encoded in the read API.

**For Agents**
Look at `ReadOpts`, `RpcFailoverClient.read/readContract`, and the EVM adapter helper methods. Preserve the current emitted consumer keys, but centralize attribution policy in one canonical read descriptor/API, or make consumer IDs explicit with a deliberate opt-out. Add/adjust focused tests proving helper and direct read paths follow the same documented policy.

File: packages/chain/src/evm-adapter-base.ts [hunk 6]
```diff
@@ -966,7 +968,10 @@ export class EVMChainAdapterBase {
     fn: (c: Contract) => Promise<T>,
     opts?: ReadOpts,
   ): Promise<T> {
-    return this.rpcFailover.readContract(label, contract, fn, opts);
+    return this.rpcFailover.readContract(label, contract, fn, {
+      ...opts,
+      rpcUsageConsumer: opts?.rpcUsageConsumer ?? label,
```
Comment (otReviewAgent) [inline packages/chain/src/evm-adapter-base.ts:973]: 🟡 Issue: `readContractWith` attribution is not exercised through the real transport path

**What's wrong**
Several production contract-view reads use `readContractWith`, and this diff changes that wrapper’s telemetry contract. The current tests cover adjacent paths, but a regression in this wrapper’s fallback to the label would leave those reads unattributed while the new suite still stays green.

**Example**
A failing-test sketch: call `a.readContractWith(contract, 'unit.contract.with', (c) => c.value())` against the loopback RPC, then assert `usage.ethCallByConsumer['unit.contract.with'] === primary.hits('eth_call') + backup.hits('eth_call')`. If line 973 stopped defaulting to the label, the current new tests would still pass via `readContract` and `readProvider`.

**Suggested direction**
Add a dedicated `readContractWith` attribution test, mirroring the existing `readContract` loopback test but using the wrapper that now owns this fallback logic.

**Confidence note**
I found existing telemetry tests that call `readContractWith`, but they use fake contracts/functions that do not exercise the counting provider path, so they do not verify the new consumer attribution behavior for this wrapper.

**For Agents**
Add a loopback transport test in `packages/chain/test/rpc-usage.unit.test.ts` that uses `EVMChainAdapter.readContractWith` with a real `Contract` view call. Assert both aggregate `byMethod.eth_call` and `ethCallByConsumer[label]` match raw loopback `eth_call` hits. Keep the existing `readContract` and `readProvider` tests intact.

File: packages/chain/src/rpc-usage.ts (hunk 3 above)
Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts:138]: Addressed in 6fdf700c3. `RpcUsageWindow` is source-compatible again at the public drain boundary: `ethCallByConsumer` is optional so aggregate-only external `RpcUsageDrainable` / `ChainAdapter` implementations still type-check. Package-owned concrete windows now use `NormalizedRpcUsageWindow`, and the new test proves a legacy aggregate-only drain source is assignable and normalized by `mergeRpcUsageWindows()`.

Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts:138]: Addressed in 6fdf700c3 as far as this can be reconciled with the red source-compatibility requirement. The loose `RpcUsageWindowInput` type and package-index normalizer export are gone. Package-owned concrete windows are modeled as `NormalizedRpcUsageWindow`, while `RpcUsageWindow` remains the source-compatible public drain shape. The CLI formatter now owns its local legacy-window normalization edge; `mergeRpcUsageWindows()` still normalizes public drain inputs intentionally so daemon composite drains remain rollout-compatible.

Comment (Jurij89) [reply packages/chain/src/rpc-usage.ts:138]: Addressed in 6fdf700c3. Added `drains eth_call consumer attribution as a per-window delta`, which records an attributed `eth_call`, drains it, then records only `eth_getLogs` and verifies the second window has `ethCallByConsumer: {}` while `lifetimeTotal` stays monotonic.

File: packages/chain/src/evm-adapter-base.ts (hunk 5 above)
Comment (Jurij89) [reply packages/chain/src/evm-adapter-base.ts:989]: Validated, but I am treating this as a follow-up rather than another in-PR API reshaping round. The current PR now has explicit `ReadOpts.rpcUsageConsumer`, preserves existing emitted consumer keys through adapter facades, and has coverage for readProvider/readContract/readContractWith, same-endpoint retry, and failover. A richer read descriptor / required attribution policy would touch the broader failover API and many call sites, so I filed #1490 for that cleanup instead of extending this attribution PR again.

File: packages/chain/src/evm-adapter-base.ts (hunk 6 above)
Comment (Jurij89) [reply packages/chain/src/evm-adapter-base.ts:973]: Addressed in 6fdf700c3. Added `attributes readContractWith eth_call attempts through the real transport path`, which drives `EVMChainAdapter.readContractWith()` through a real loopback ethers `Contract` view with primary 429 + backup success, then asserts both aggregate `byMethod.eth_call` and `ethCallByConsumer['unit.contract.with']` match raw loopback `eth_call` hits.

File: packages/chain/src/evm-adapter-base.ts (hunk 5 above)
Comment (otReviewAgent) [reply packages/chain/src/evm-adapter-base.ts:989]: 🟡 Issue: Do not make human read labels the implicit telemetry contract

**What's wrong**
This overloads one string with two responsibilities: a human failover/span label and a stable low-cardinality daemon-log key. That makes the telemetry contract fragile because unrelated readability changes to labels become externally visible log-shape changes. It also contradicts the new `ReadOpts` boundary, which is trying to keep the consumer key explicit.

**Example**
A future cleanup that renames a span/failover label from `token.allowance` to `token allowance funding read` would also change the emitted `rpc_usage_by_consumer` key after normalization, even though the telemetry consumer contract was not intentionally changed.

**Suggested direction**
Split the two concepts at the adapter boundary. Either require an explicit `rpcUsageConsumer` for attributed reads, or introduce a small typed/read-descriptor abstraction that carries both `label` and stable `consumer` fields. Avoid defaulting the consumer from the display/span label.

**For Agents**
Look at `EVMChainAdapterBase.readContract`, `readContractWith`, and `readProvider`. Preserve the current attribution values, but separate the span/failover label from the consumer key: use explicit read descriptors, consumer constants, or an opt-in helper so telemetry keys are owned deliberately instead of inherited from human labels. Prove existing attributed reads still emit the same consumer values.

File: packages/cli/src/daemon/rpc-usage-log.ts [hunk 7]
```diff
@@ -7,15 +7,26 @@
...
+  type RpcUsageDrainable,
+  type RpcUsageWindow,
+} from '@origintrail-official/dkg-chain';
 
 /** logfmt-token safety: methods/chain ids are self-generated, but never emit a token that could break parsing. */
 function safeToken(value: string, fallback: string): string {
   return /^[A-Za-z0-9_.:-]{1,64}$/.test(value) ? value : fallback;
 }
 
+function normalizeRpcUsageWindow(usage: RpcUsageWindow): RpcUsageWindow & { ethCallByConsumer: Record<string, number> } {
```
Comment (otReviewAgent) [inline packages/cli/src/daemon/rpc-usage-log.ts]: 🟡 Issue: Reuse the chain package’s window normalizer instead of duplicating it in the CLI

**What's wrong**
The PR introduces a canonical normalization function in the chain model, then hand-rolls the same compatibility shim in the daemon formatter. That splits ownership of the `RpcUsageWindow` shape across packages and makes future telemetry dimensions more likely to drift between merge/drain behavior and log formatting.

**Example**
If the next window shape adds another optional diagnostic map, `mergeRpcUsageWindows` can normalize it in the chain package while `formatRpcUsageLines` keeps silently using the older local CLI normalization. The formatter then becomes the second place that must remember every legacy-window compatibility rule.

**Suggested direction**
Make normalization a single owned boundary in `@origintrail-official/dkg-chain` and have the daemon formatter consume that helper or a `NormalizedRpcUsageWindow`. The CLI should format an already-normalized model, not duplicate compatibility logic for the model.

**For Agents**
Export `normalizeRpcUsageWindow` from `packages/chain/src/index.ts` or otherwise normalize before the CLI boundary, then remove the local helper in `packages/cli/src/daemon/rpc-usage-log.ts`. Keep legacy aggregate-only windows formatting exactly as they do now.

File: packages/chain/src/evm-adapter-base.ts (hunk 5 above)
Comment (Jurij89) [reply packages/chain/src/evm-adapter-base.ts:989]: Validated, and still treating this as follow-up #1490 rather than another in-PR API reshape. The current PR has explicit `ReadOpts.rpcUsageConsumer`, preserves current emitted keys, and covers the attribution paths we changed. Making labels/span names/usage consumers a single typed read descriptor is the right cleanup, but it affects the broader failover read API and should be reviewed independently instead of extending this attribution PR again.

File: packages/cli/src/daemon/rpc-usage-log.ts (hunk 7 above)
Comment (Jurij89) [reply packages/cli/src/daemon/rpc-usage-log.ts]: Addressed in a47b3179f. `normalizeRpcUsageWindow()` is now exported from `@origintrail-official/dkg-chain`, and the daemon formatter imports that canonical normalizer instead of carrying a local duplicate. This keeps source compatibility (`RpcUsageWindow.ethCallByConsumer` remains optional at the public boundary) without reintroducing the loose `RpcUsageWindowInput` export.

Verified with chain build/tsc, chain RPC usage unit tests, CLI formatter tests, CLI tsc, and `git diff --check`.

File: packages/chain/src/evm-adapter-base.ts (hunk 5 above)
Comment (otReviewAgent) [reply packages/chain/src/evm-adapter-base.ts:989]: 🟡 Issue: Do not make human read labels double as stable telemetry consumers

**What's wrong**
The new option is documented as an explicit low-cardinality telemetry consumer, but the adapter base immediately falls back to the existing human `label`. That couples failover/span wording to daemon log identity and spreads the telemetry contract across dozens of unrelated call sites. The escape hatch exists, but maintainers must notice and use it every time they edit a label.

**Example**
`readContractWith(token, 'allowance visibility poll', ...)` now emits attribution under `consumer=allowance_visibility_poll` by default. A future rename of that human failover/span label would silently rename the telemetry consumer too.

**Suggested direction**
Introduce a dedicated stable consumer key at the canonical read-helper boundary, or model reads as `{ label, rpcUsageConsumer }` where the consumer is intentionally chosen. Avoid the blanket `?? label` fallback for all reads unless the label is formally promoted to the stable telemetry contract.

**For Agents**
In `packages/chain/src/evm-adapter-base.ts`, separate human read labels from stable consumer attribution. Preserve existing read behavior, but make the consumer key explicit in the domain read definition or a small typed read descriptor, and update tests to prove a label-only rename cannot change `rpc_usage_by_consumer`.

Comment (Jurij89) [reply packages/chain/src/evm-adapter-base.ts:989]: Validated again, and I still think this should stay as follow-up #1490 rather than another PR #1487 code round. The concern is real design hygiene, but it is the broader read-descriptor/API-shape cleanup: removing the `?? label` default properly means choosing stable consumer IDs across many domain reads and reviewing that as an intentional telemetry taxonomy change.

For this attribution PR, the current behavior is explicit enough to measure `eth_call` consumers now: `ReadOpts.rpcUsageConsumer` exists, existing emitted keys are preserved deliberately, and the changed readProvider/readContract/readContractWith/retry/failover paths are covered. I do not see a correctness bug or broken public contract here, so I am leaving the deeper API cleanup to #1490 to avoid more churn before we get measurement data.
